attestation-tool/src/ecdsa.rs

```diff
@@ -0,0 +1,369 @@
+use std::{ffi::CString, str};
+
+use oasis_core_runtime::common::sgx::pcs::{QEIdentity, TCBInfo, TCBLevel, TCBStatus};
+
+use aesm_client::AesmClient;
+use anyhow::{anyhow, Result};
+use byteorder::{ByteOrder, LittleEndian};
+use dcap_ql::quote::Qe3CertDataPckCertChain;
+use dcap_ql::quote::{Quote, Quote3SignatureEcdsaP256};
+use mbedtls::{alloc::List as MbedtlsList, x509::certificate::Certificate};
+use rustc_hex::FromHex;
+use sgx_isa::{Report, Targetinfo};
+use sgxs_loaders::isgx::Device as IsgxDevice;
+
+// Intel's PCS signing root certificate.
+const PCS_TRUST_ROOT_CERT: &str = r#"-----BEGIN CERTIFICATE-----
+MIICjzCCAjSgAwIBAgIUImUM1lqdNInzg7SVUr9QGzknBqwwCgYIKoZIzj0EAwIw
+aDEaMBgGA1UEAwwRSW50ZWwgU0dYIFJvb3QgQ0ExGjAYBgNVBAoMEUludGVsIENv
+cnBvcmF0aW9uMRQwEgYDVQQHDAtTYW50YSBDbGFyYTELMAkGA1UECAwCQ0ExCzAJ
+BgNVBAYTAlVTMB4XDTE4MDUyMTEwNDUxMFoXDTQ5MTIzMTIzNTk1OVowaDEaMBgG
+A1UEAwwRSW50ZWwgU0dYIFJvb3QgQ0ExGjAYBgNVBAoMEUludGVsIENvcnBvcmF0
+aW9uMRQwEgYDVQQHDAtTYW50YSBDbGFyYTELMAkGA1UECAwCQ0ExCzAJBgNVBAYT
+AlVTMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEC6nEwMDIYZOj/iPWsCzaEKi7
+1OiOSLRFhWGjbnBVJfVnkY4u3IjkDYYL0MxO4mqsyYjlBalTVYxFP2sJBK5zlKOB
+uzCBuDAfBgNVHSMEGDAWgBQiZQzWWp00ifODtJVSv1AbOScGrDBSBgNVHR8ESzBJ
+MEegRaBDhkFodHRwczovL2NlcnRpZmljYXRlcy50cnVzdGVkc2VydmljZXMuaW50
+ZWwuY29tL0ludGVsU0dYUm9vdENBLmRlcjAdBgNVHQ4EFgQUImUM1lqdNInzg7SV
+Ur9QGzknBqwwDgYDVR0PAQH/BAQDAgEGMBIGA1UdEwEB/wQIMAYBAf8CAQEwCgYI
+KoZIzj0EAwIDSQAwRgIhAOW/5QkR+S9CiSDcNoowLuPRLsWGf/Yi7GSX94BgwTwg
+AiEA4J0lrHoMs+Xo5o/sX6O9QWxHRAvZUGOdRQ7cvqRXaqI=
+-----END CERTIFICATE-----"#;
+lazy_static::lazy_static! {
+    static ref PCS_TRUST_ROOT: MbedtlsList<Certificate> = {
+        let mut cert_chain = MbedtlsList::new();
+        let raw_cert = CString::new(PCS_TRUST_ROOT_CERT.as_bytes()).unwrap();
+        let cert = Certificate::from_pem(raw_cert.as_bytes_with_nul()).unwrap();
+        cert_chain.push(cert);
+
+        cert_chain
+    };
+}
+
+// OIDs for PCK X509 certificate extensions.
+const PCK_SGX_EXTENSIONS_OID: &[u64] = &[1, 2, 840, 113741, 1, 13, 1];
+const PCK_SGX_EXTENSIONS_FMSPC_OID: &[u64] = &[1, 2, 840, 113741, 1, 13, 1, 4];
+const PCK_SGX_EXTENSIONS_TCB_OID: &[u64] = &[1, 2, 840, 113741, 1, 13, 1, 2];
+
+// TCB url. // TODO: configurable.
+const TCB_URL: &str = "https://api.trustedservices.intel.com/sgx/certification/v4/tcb";
+const QE_IDENTITY_URL: &str =
+    "https://api.trustedservices.intel.com/sgx/certification/v4/qe/identity";
+
+#[derive(Clone, Debug, Default, serde::Deserialize)]
+/// Response from TCBInfo API.
+struct TCBInfoResponse {
+    #[serde(rename = "tcbInfo")]
+    pub tcb_info: TCBInfo,
+}
+
+// Taken from TCBLevel.matches (which is a private function).
```

Comment:
Let's update Oasis Core to expose this so we can reuse it here in a later release.